### PR TITLE
Dynamic Filter network

### PR DESCRIPTION
**Pull request by PatWie**
Dynamic Filter Networks <https://arxiv.org/abs/1605.09673>
by Bert De Brabandere et al.

This also demonstrates `monitors.put_image` from #239, which gives a nice slider for the `filter_export` summary in the newest TF version.

### CODE REVIEW

**Comment by ppwwyyxx** [comment]
Looks like this is section4.1 of the paper. 
Were you able to see the network learning expected filters? If it's successful it would be a good example to add.

**Comment by PatWie** [comment]
if you run this network, you can use tensorboard and will see these filters after a while. Tensorboard will show (OnlineTensorboardCallback) both the expected and learned filters.

**Comment by ppwwyyxx** [comment]
Since they did open source their code https://github.com/dbbert/dfn/blob/master/experiment_steerableFilter_tensorflow.ipynb, I guess it would be better to follow the network & hyperparam rather than making things up.
Plus they seem to have a more efficient dynamic conv implementation.

**Comment by PatWie** [comment]
I changed the conv2d to fullyconnected layers and get a faster convergence:

tensorboard_dfn

**Comment by PatWie** [comment]
I don't think, that their "dynamic conv implementation" is more efficient (in terms of speed *and* memory). Rather than  using their explicit im2col, this implementation uses a highly optimized call of TF. 
Further, TensorFlow uses the cuda streams, so this would also hide the latency from data transfers between kernel calls from the `[... zip(...)]`.

**Comment by ppwwyyxx** [comment]
Do you know why they get to a very small loss with only 490 steps (and their loss is summed over the image and averaged over batch). Maybe because their dataset is easier?
Anyway the current result you showed is pretty good.

**Comment by PatWie** [comment]
I have collected a large image data set, where I simply rejected images without any prominent edges. Using this custom dataset helps a lot to converge faster.
So, I would guess it depends heavily on the training data, e.g. try MNIST.

File: examples/DynamicFilterNetwork/steering-filter.py [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+
+        Returns:
+            learned filter as [B, k, k]
+        """
+        with argscope(LeakyReLU, alpha=0.2), \
+                argscope(FullyConnected, nl=LeakyReLU):
+            net = FullyConnected('fc1', theta, 64)
+            net = FullyConnected('fc2', net, 128)
+
+        pred_filter = FullyConnected('conv3', net, kernel_shape ** 2, nl=tf.identity)
```
Comment (ppwwyyxx) [inline examples/DynamicFilterNetwork/steering-filter.py]: Probably 'conv3' isn't the desired name?

File: examples/DynamicFilterNetwork/steering-filter.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+        gt_image = tf.reshape(gt_image, [BATCH, SHAPE, SHAPE, 1])
+
+        pred_filter = self._parameter_net(theta)
+        pred_image = dynamic_conv_filter('dfn', image, pred_filter, 1, kernel_size)
+
+        with tf.name_scope('visualization'):
+            def n(x):
+                x = x - tf.reduce_min(x)
+                x = x / tf.reduce_max(x)
+                return x
```
Comment (ppwwyyxx) [inline examples/DynamicFilterNetwork/steering-filter.py]: tf.summary.image will do some kind of normalization when input is float. Maybe this is not necessary?

File: examples/DynamicFilterNetwork/steering-filter.py [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+    rsl = tf.concat(rsl, axis=0, name='output')
+    return rsl
+
+
+class OnlineTensorboardExport(Callback):
+    """Show learned filters for specific thetas in TensorBoard.
+    """
+    def __init__(self):
+        # generate 32 filters (8 different, 4 times repeated)
+        self.theta = np.array(range(8) * 4) / 8. * 2 * np.pi
```
Comment (ppwwyyxx) [inline examples/DynamicFilterNetwork/steering-filter.py]: `range(8)*4` is not python3 compatible.

File: examples/DynamicFilterNetwork/steering-filter.py [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+            theta = self.rng.uniform(0, 2 * np.pi)
+            filtered_image, used_filter = ThetaImages.filter_with_theta(image[0], theta)
+            yield [theta, image[0], filtered_image, used_filter]
+
+
+def get_data():
+    # probably not the best dataset
+    ds = dataset.BSDS500('train', shuffle=True)
+    # rgb2gray
+    ds = MapData(ds, lambda dp: [np.dot(dp[0], [0.299, 0.587, 0.114])[:, :]])
```
Comment (ppwwyyxx) [inline examples/DynamicFilterNetwork/steering-filter.py]: You added `Grayscale ` augmentor but is not using it ..

File: examples/DynamicFilterNetwork/steering-filter.py (hunk 1 above)
Comment (PatWie) [inline examples/DynamicFilterNetwork/steering-filter.py]: Why do you not just give my creativity some freedom? ;-)

**Comment by PatWie** [comment]
I changed these lines and also partly tested python3 compatibility.
